### PR TITLE
Support for exporting to YAML format

### LINKED ISSUE #6974: Alternative serialization format as core data file (e.g. YAML in addition to BibTeX)
Alternative serialization format support (e.g. YAML in addition to BibTeX)

Probably many use cases where `JabRef`'s data file is used include `Pandoc`. Although it is common to use `BibTeX` as input into `pandoc-citeproc`, `YAML` serialized `BibTeX` (citeproc YAML) is also an option. When using `Pandoc`'s flavour of `markdown`, `YAML` is also the format of the metadata block.

It would be fantastic if `JabRef` could read and write `YAML` serialized `BibTeX`. In addition YAML is also less cumbersome to deal with when one needs to do some manual editing of the data file.

Furthermore by supporting `YAML` there is also a possibility of directly copy pasting selected references from `JabRef` data file into the `markdown`'s metadata segment (in addition to the possibility of using external references file exclusively).

Bellow is an example of currently the only `JabRef`'s option and under it what would be great to see `JabRef` also supporting.

Original in `BibTeX`:

```bib
% Encoding: UTF-8
@Misc{id_doc,
    author = {{John Doe}},
    title  = {{Doc Title}},
    date   = {2020-10-03},
    url    = {http://www.example.com}
}
```

Converted to YAML that is recognized by `Pandoc` / `pandoc-citeproc`:

```yaml
---
references:
- id: id_doc
  type: no-type
  author:
  - literal: John Doe
  title: Doc Title
  issued: [[2020,10,3]]
  URL: http://www.example.com
...
```

### PR DESCRIPTION
**Pull request by joethei**
- Support for exporting to YAML , fixes #6974 
- Fixed deleteBlankLines parameter in TemplateExporter not working.
- Enabled deleteBlankLines for the RIS Exporter, which used to have that enabled.

- [x] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Review by koppor — CHANGES_REQUESTED** [review]
Looks good. Some nitpicks remaining. The empty line fix should have gone in a separate PR as discussions on regular expressions typically take more time. Let's see if it goes well, here.

Could you also add a test case for the example at https://github.com/JabRef/jabref/issues/6974? I think, there could be issues at the date export. - Maybe good for a follow-jup PR?

File: src/main/resources/resource/layout/yaml.layout [hunk 1]
```diff
@@ -0,0 +1,14 @@
+- id: \citationkey
+\begin{type}  type: \type\end{type}
+\begin{author}
+  author:
+  - literal: "\author"
+\end{author}
+\begin{title}  title: "\title"\end{title}
+\begin{shorttitle}  title-short: "\shorttitle"\end{shorttitle}
+\begin{date}  issued: \date\end{date}
```
[the remark this answers is not in the document]
Comment (joethei) [reply src/main/resources/resource/layout/yaml.layout:9]: no that is not the shown date format, Pandoc accepts a number of different formats for dates.
Using pandoc to convert a .bib file to .yaml also outputs yyyy-mm-dd.

File: src/main/java/org/jabref/logic/exporter/TemplateExporter.java [hunk 2]
```diff
@@ -278,8 +269,12 @@ public void export(final BibDatabaseContext databaseContext, final Path file,
                 // Write the entry
                 if (layout != null) {
                     if (deleteBlankLines) {
-                        String withoutBlankLines = BLANK_LINE_MATCHER.matcher(layout.doLayout(entry, databaseContext.getDatabase())).replaceAll("");
-                        ps.write(withoutBlankLines);
+                        String[] lines = layout.doLayout(entry, databaseContext.getDatabase()).split("\n");
```
[the remark this answers is not in the document]
Comment (Siedlerchr) [reply src/main/java/org/jabref/logic/exporter/TemplateExporter.java:272]: Maybe it should check both? We don't know which line endings the user has in the template files (if modified)

File: src/main/resources/resource/layout/yaml.layout (hunk 1 above)
[the remark this answers is not in the document]
Comment (koppor) [reply src/main/resources/resource/layout/yaml.layout:9]: Is it possible to add the test bib from #6974 or is that difficult?

**Comment by koppor** [comment]
I see the unit tests failing:

```
  Test exportCorrectContent(Path) FAILED

  org.opentest4j.AssertionFailedError: expected: <[---, references:, - id: test,   author:,   - literal: "Test Author",   title: "Test Title",   issued: 2020,   url: http://example.com, ---]> but was: <[---, references:, - id: test  author:  - literal: "Test Author"  title: "Test Title"  issued: 2020  url: http://example.com---]>
      at org.jabref.logic.exporter.YamlExporterTest.exportCorrectContent(YamlExporterTest.java:77)
```

Could you please check what's going wrong there?

**Comment by joethei** [comment]
> 
> 
> I see the unit tests failing:
> 
> ```
>   Test exportCorrectContent(Path) FAILED
> 
>   org.opentest4j.AssertionFailedError: expected: <[---, references:, - id: test,   author:,   - literal: "Test Author",   title: "Test Title",   issued: 2020,   url: http://example.com, ---]> but was: <[---, references:, - id: test  author:  - literal: "Test Author"  title: "Test Title"  issued: 2020  url: http://example.com---]>
>       at org.jabref.logic.exporter.YamlExporterTest.exportCorrectContent(YamlExporterTest.java:77)
> ```
> 
> Could you please check what's going wrong there?

I already fixed that issue in my local branch, currently testing the blank line deletion.

**Review by Siedlerchr — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -52,8 +53,10 @@ inserting new citations in a OpenOffic/LibreOffice document. [#6957](https://git
 - We fixed an issue where it was impossible to connect to OpenOffice/LibreOffice on Mac OSX. [#6970](https://github.com/JabRef/jabref/pull/6970)
 - We fixed an issue with the python script used by browser plugins that failed to locate JabRef if not installed in its default location. [#6963](https://github.com/JabRef/jabref/pull/6963/files)
 - We fixed an issue where identity column header had incorrect foreground color in the  Dark theme. [#6796](https://github.com/JabRef/jabref/issues/6796)
+- We fixed an issue where the RIS exporter added extra blank lines.[#7007](https://github.com/JabRef/jabref/pull/7007/files)
 - We fixed an issue where clicking on Collapse All button in the Search for Unlinked Local Files expanded the directory structure erroneously [#6848](https://github.com/JabRef/jabref/issues/6848)
 
+
```
Comment (calixtus) [inline CHANGELOG.md:59]: Checkstyle linting error. Only one white line allowed

File: src/main/java/org/jabref/logic/exporter/TemplateExporter.java [hunk 4]
```diff
@@ -112,27 +102,33 @@ public TemplateExporter(String displayName, String consoleName, String lfFileNam
...
      * @param extension         Should contain the . (for instance .txt).
      * @param layoutPreferences Preferences for layout
      * @param savePreferences   Preferences for saving
-     * @param deleteBlankLines  If blank lines should be remove (default: false)
+     * @param deleteBlankLines  If blank lines should be removed (default: false)
      */
-    public TemplateExporter(String consoleName, String lfFileName, String directory, StandardFileType extension, LayoutFormatterPreferences layoutPreferences, SavePreferences savePreferences, boolean deleteBlankLines) {
-        this(consoleName, consoleName, lfFileName, directory, extension, layoutPreferences, savePreferences);
+    public TemplateExporter(String displayName, String consoleName, String lfFileName, String directory, FileType extension,
+                            LayoutFormatterPreferences layoutPreferences, SavePreferences savePreferences, boolean deleteBlankLines) {
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/exporter/TemplateExporter.java:117]: Can you make `boolean deleteBlankLines` an enum? `KEEP_BLANKS` or `DELETE_BLANKS` is way more intuitive as `true` and `false` at the caller side.

File: src/main/java/org/jabref/logic/exporter/TemplateExporter.java [hunk 5]
```diff
@@ -112,27 +102,33 @@ public TemplateExporter(String displayName, String consoleName, String lfFileNam
...
+        this.layoutPreferences = layoutPreferences;
+        this.savePreferences = savePreferences;
         this.deleteBlankLines = deleteBlankLines;
     }
 
     /**
-     * Indicate whether this is a custom export. A custom export looks for its
-     * layout files using a normal file path, while a built-in export looks in
-     * the classpath.
+     * Indicate whether this is a custom export. A custom export looks for its layout files using a normal file path, while a built-in export looks in the classpath.
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/exporter/TemplateExporter.java:131]: Please wrap the comments after 120 characters, to keep it readable in github

**Review by calixtus — CHANGES_REQUESTED** [review]
Got three nitpicks.
Thanks for the good work so far.

**Review by calixtus — APPROVED** [review]
LGTM! Thank you!

File: src/main/resources/resource/layout/yaml.layout (hunk 1 above)
[the remark this answers is not in the document]
Comment (joethei) [reply src/main/resources/resource/layout/yaml.layout:9]: adding the exact yaml file as a test case ist not difficult, but i dont think it ist necessary as pandoc also uses yyyy-mm-dd as a default. I did a a testcase that uses all those fields.

File: src/main/java/org/jabref/logic/exporter/BlankLineBehaviour.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package org.jabref.logic.exporter;
+
+/**
+ this enum represents the behaviour for blank lines in {@link TemplateExporter}
```
Comment (koppor) [inline src/main/java/org/jabref/logic/exporter/BlankLineBehaviour.java:4]: Full JavaDoc 😇 

```suggestion
 * This enum represents the behaviour for blank lines in {@link TemplateExporter}
```

File: src/main/java/org/jabref/logic/exporter/TemplateExporter.java [hunk 7]
```diff
@@ -33,20 +33,14 @@
...
-     * ?m activates "multimode", which makes ^ match line starts/ends.
-     * \\s simply marks any whitespace character
-     */
-    private static final Pattern BLANK_LINE_MATCHER = Pattern.compile("(?m)^\\s");
     private static final String LAYOUT_EXTENSION = ".layout";
     private static final String FORMATTERS_EXTENSION = ".formatters";
     private static final String BEGIN_INFIX = ".begin";
     private static final String END_INFIX = ".end";
 
+
```
Comment (koppor) [inline src/main/java/org/jabref/logic/exporter/TemplateExporter.java:43]: Please do not add two empty lines - I wonder why checkstyle did not complain

File: src/main/java/org/jabref/logic/layout/format/CSLType.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.jabref.logic.layout.LayoutFormatter;
+import org.jabref.model.entry.types.StandardEntryType;
+
+public class CSLType implements LayoutFormatter {
+
+    @Override
+    public String format(String value) {
+        Map<String, String> map = Map.of(StandardEntryType.Article.getDisplayName(), "article",
+                StandardEntryType.Book.getDisplayName(), "book",
```
Comment (koppor) [inline src/main/java/org/jabref/logic/layout/format/CSLType.java:13]: A simple  switch was not possible

```java
  return switch(value) {
     case StandardEntryType.Article.getDisplayName() -> "article",
     ...
     case StandardEntryType.Conference.getDisplayName() -> "paper-conference",
     ...
     default -> "no-type";
  }
```

**Review by koppor — CHANGES_REQUESTED** [review]
Sorry for delay. Hope, you are still in the context. I have some micro comments...

**Comment by koppor** [comment]
Thank you for the patience @joethei

**Pull request merged**